### PR TITLE
Allow download with duplicate filenames

### PR DESCRIPTION
**Pull request by Zenodeon**
PR Fixes files not being saved when a file with the same name exists in the download folder.

It is fixed by creating a new temp folder (named by the unique identifier) in the download folder for each video.

Then all the embedding and processing are done in this folder and once done, the final file (be it mp4, opus, or whatever the user has requested) will be moved out of the temp folder.

While moving it out, it checks for any pre-existing file with the same name, and if it exists, it adds "(n)" to the file name where n represents the duplicate file name count.

After Moving the processed file, the temp folder is forced Deleted.

### CODE REVIEW

**Comment by StefanLobbenmeier** [comment]
Before we merge this, I want to check if yt-dlp can do this for us.

This is the output when you download a video for the second time:

```
[youtube] Extracting URL: https://www.youtube.com/watch?v=7vZ020mPm3k
[youtube] 7vZ020mPm3k: Downloading webpage
[youtube] 7vZ020mPm3k: Downloading android player API JSON
[info] 7vZ020mPm3k: Downloading 1 format(s): 597+140
[download] C:\Users\Stefan\Downloads\  \DAEGHO - ENTER THE FUEGO-(144p12).mp4 has already been downloaded
[Metadata] Adding metadata to "C:\Users\Stefan\Downloads\  \DAEGHO - ENTER THE FUEGO-(144p12).mp4"
Download finished
```

**Comment by StefanLobbenmeier** [comment]
Checked through yt-dlp source code, and I cannot see an additional condition to not skip the download. So you are right, we will have to implement it ourselves

**Review by StefanLobbenmeier — CHANGES_REQUESTED** [review]

**Comment by StefanLobbenmeier** [comment]
Will also test this on my PC now

File: modules/types/Video.js
Comment (StefanLobbenmeier) [reply modules/types/Video.js:58]: ahh no, one says download, the other merger in the tag. Might also fix some issue we are currently having. Ideally we would also use -J, already implemented --progress-template like here: 70265863

but for now this is fine

File: modules/download/DownloadQuery.js
Comment (StefanLobbenmeier) [reply modules/download/DownloadQuery.js:28]: Unfortunately not, but dont worry about that

**Comment by StefanLobbenmeier** [comment]
Local testing worked just fine. The biggest issue I have with this is that this behaviour might not be wanted by everyone.

For example it breaks resumability, since the files are in the temp folder. So we need some kind of toggle to turn this feature off or on. Would you also be able to add that?

File: modules/download/DownloadQuery.js
Comment (Zenodeon) [reply modules/download/DownloadQuery.js:154]: Ah, did not see the if statement up there.

It is not necessarily needed, as the unmerge files are saved in the temp folder, so it is no longer needed for yt-dlp to delete them, as when the video is done processing, I am forcing deleting the temp folder which will delete those files as well.

I did not think about this at the time of writing the code, but I think this is a better way as it does not involve two threads to handle contents in that folder (1 thread from yt-dlp and another from ours).

so maybe we can reuse keepUnmerged as KeepTempFolder.

**Comment by Zenodeon** [comment]
I like to keep this as a core feature as we can't control video names and it can be the same sometimes. Which won't be saved and the user might not notice even. 

If Resumability is the only issue, I will take a look at that feature and make it compatible with the temp folder or else i will go with your plan of making it as an option.

File: modules/download/DownloadQuery.js
Comment (Zenodeon) [reply modules/download/DownloadQuery.js:28]: yea, for some reason, The App does not store/get the playlist data.

**Comment by Zenodeon** [comment]
What is the Resumability feature btw? I thought it was being able to pause the download and start the download from where it stopped.

File: modules/download/DownloadQuery.js
Comment (Zenodeon) [reply modules/download/DownloadQuery.js:154]: or rename the temp folder to the video name.

**Comment by StefanLobbenmeier** [comment]
Currently you can start a download, close the app, reopen it and download the file again. Yt-dlp will see that it has partially downloaded the file and resume where it left off. This is now broken, because we put the files in a new temp folder, invisible to yt-dlp

i also consider it a feature that downloading a file a second time instantly completes, because the file is already present. 

In your case, with different videos having the same filename one might also consider it a feature that only one of them is kept, they are the same song after all. If a user does not want it, right now they can include the video id in the output template to make sure different video ids are saved to separate files. That would also be the easiest solution for your use case.


If you really want to have enumeration when the file already exists, i would propose to add a special template in the output template. For example if the output template contains %s(dup_index), we can do the logic checking if the file already exists and add a number to the file name. Also saves us from building any additional ui

**Comment by Zenodeon** [comment]
yea that's fair, I will add an option below the "Keep unmerged files" option for this feature.

**Comment by Zenodeon** [comment]
Resumability Works with the folder system.

While Avoid Failing To Save Duplicate File Name Option won't but it will never have problems with a duplicate file name.

**Comment by Zenodeon** [comment]
if you can help add a note below that option stating that Resumability function won't work while Avoid Failing To Save Duplicate File Name Option is active will be great.

File: modules/types/Video.js
Comment (StefanLobbenmeier) [inline modules/types/Video.js:155]: this does not work, yt-dlp supports all kinds of urls

File: modules/download/DownloadQuery.js
Comment (StefanLobbenmeier) [inline modules/download/DownloadQuery.js:30]: since this is the default case, we also do not have to create any temp folder. lets just use the previous logic here

File: modules/download/DownloadQuery.js
Comment (StefanLobbenmeier) [inline modules/download/DownloadQuery.js:159]:
```suggestion
        if(this.environment.settings.keepUnmerged || this.environment.settings.avoidFailingToSaveDuplicateFileName) args.push('--keep-video');
```

**Comment by StefanLobbenmeier** [comment]
> if you can help add a note below that option stating that Resumability function won't work while Avoid Failing To Save Duplicate File Name Option is active will be great. 

since we have this as an option now, I think we are ok without it. Otherwise I will just add a tooltip to it with the details

File: renderer/renderer.html
Comment (StefanLobbenmeier) [inline renderer/renderer.html:390]: I dont quite like the name for this, can we call it "Enumate Duplicate Filenames" instead?

**Review by StefanLobbenmeier — CHANGES_REQUESTED** [review]

File: modules/types/Video.js
Comment (StefanLobbenmeier) [reply modules/types/Video.js:155]: (method can be removed if we use default logic)

**Comment by StefanLobbenmeier** [comment]
I am not allowed to edit your branch, so I also created a branch in this repo

File: renderer/renderer.html
Comment (StefanLobbenmeier) [reply renderer/renderer.html:390]: 6e287ec6

Comment (Zenodeon) [reply renderer/renderer.html:390]: Now all files will be Enumated if there is a duplicate file name in the folder. so this name wont make sense.

I named it like this as making this active will make the Resumability function inaccessible.

File: modules/download/DownloadQuery.js
Comment (Zenodeon) [inline modules/download/DownloadQuery.js:157]: Hey, KeepUnmerged is now handled at line 217 in the old code and line 215 in this change.

File: modules/download/DownloadQuery.js
Comment (Zenodeon) [reply modules/download/DownloadQuery.js:30]: what do you mean by default case?

File: modules/types/Video.js
Comment (Zenodeon) [reply modules/types/Video.js:155]: It worked for me, i also tested with several videos and yt-dlp gave the url only in https ones.

If you say it will give different ones, i will make a bit dynamic one.

File: modules/download/DownloadQuery.js
Comment (Zenodeon) [reply modules/download/DownloadQuery.js:30]: The one using 'this.video.identifier' will never fail to save a duplicate file name but Resumability won't work, while the one using 'this.video.GetURLID()' has an edge case of failing to save a duplicate name (like when the user is downloading the same video more than once) but it lets you use Resumability.

**Comment by Zenodeon** [comment]
Please check commit [7a5adc3](https://github.com/StefanLobbenmeier/youtube-dl-gui/pull/22/commits/7a5adc3302052f4cfdd8ecd95fb82bfdfb3d8077)

Isolating my feature from the standard/core system.

while at it, found a new bug, will create a new issue.

**Comment by Zenodeon** [comment]
nvm, will do it here as it is only relevant to my feature only.

Screenshot_2658

Any idea why [Merger] or [Metadata] dont report the file name correctly?

**Comment by Zenodeon** [comment]
Here is the video if you need to. 

Excuse the video content. 
https://www.youtube.com/watch?v=PYeZX7X9tB0

**Comment by StefanLobbenmeier** [comment]
> nvm, will do it here as it is relevant to my feature only.
> 
> Screenshot_2658
> 
> Any idea why [Merger] or [Metadata] don't report the file name correctly?
> 
> Only happens when the video title has char like "/", so far from testing.

the \u29f8 is unicode encoded it seems: https://unicode-table.com/de/29F8/

File: renderer/renderer.html
Comment (StefanLobbenmeier) [reply renderer/renderer.html:390]: I am mostly unhappy with that naming since it makes the default case a "failure", which is not the case.

> Now all files will be Enumated if there is a duplicate file name in the folder. so this name wont make sense.

Surely we only add the (1) if the file already exists, not add a number to every download, no?

**Comment by Zenodeon** [comment]
Yea ik, but it works fine for filename with extention. And yt-dlp saves the file with that name, even reports ot right but when it embed the metadata or mergers, it reports the file name wrong.

File: renderer/renderer.html
Comment (Zenodeon) [reply renderer/renderer.html:390]: > I am mostly unhappy with that naming since it makes the default case a "failure", which is not the case.
I mean, it is the case, what's why I am doing this lol. Few of my videos failed to download from a playlist of 300 videos, took me a while to figure them out. 

The logs were not opening too with that playlist, another bug because the log was too big?

>Surely we only add the (1) if the file already exists, not add a number to every download, no?
This is no longer the case since the latest commit. If this option is not active, noting happens. yt-dlp will save the unmerged file in the download folder and merged them in the same folder.

Comment (Zenodeon) [reply renderer/renderer.html:390]: > We can also easily add a more description on hover using the title attribute, it will look like this:

Thank you so much. Will document it in the Wiki once this PR is merged.

**Comment by Zenodeon** [comment]
I can deal with this issue with a funky code, but checking if it can be fixed from yt-dlp.

File: renderer/renderer.html
Comment (Zenodeon) [reply renderer/renderer.html]: @StefanLobbenmeier

Comment (StefanLobbenmeier) [reply renderer/renderer.html]: Sorry did not check the PR last week, but should be able to get this merged this weekend

**Comment by StefanLobbenmeier** [comment]
There are some code style issues, but I will fix them in a separate branch. For now, it looks good to me

**Comment by StefanLobbenmeier** [comment]
Reminder for the future: yt-dlp actually has this built-in and its called %(autonumber)s

**Pull request merged**
